### ISSUE TITLE
[Bigquery insert data error] table.insert_data() issue

### ISSUE THREAD

**Issue by icrtiou** [GoogleCloudPlatform/gcloud-python#1381]
I got this error while using `table.insert_data(rows)`

```
--> 246     if not value.tzinfo:
    247         value = value.replace(tzinfo=UTC)
    248     # Regardless of what timezone is on the value, convert it to UTC.

AttributeError: 'str' object has no attribute 'tzinfo'
```

The doc of `table.insert_data()` said `rows` is list of tuple, so this is my schema
```
sql_schema = [('viewer_id', 'INTEGER'),
              ('ip','STRING'),
              ('author_id', 'INTEGER'),
              ('post_id', 'STRING'),
              ('time', 'TIMESTAMP')]
``` 

And this is a test data:
```
[['25512',
  '126.107.92.215',
  '6524',
  '9f4047d3-7bf8-406c-823c-9803b4730fb9',
  '2015-12-14 18:29:01.245142'],
 ['24534',
  '87.13.200.79',
  '29096',
  'e09baadc-9647-4cde-9019-7c6846dfd662',
  '2015-12-14 18:29:01.245237']]
```

I know ` '2015-12-14 18:29:01.245237'` is a string which doesn't have `tzinfo` field. 
Do I have to convert it to Datetime object every time I try to load time data into bigquery?
Do I have better options?

Thanks

**Comment by dhermes** [comment]
@icrtiou Are you on

http://googlecloudplatform.github.io/gcloud-python/latest/

or

http://googlecloudplatform.github.io/gcloud-python/0.8.0/

Also which version do you have installed?

```
$ pip show gcloud
---
Metadata-Version: 2.0
Name: gcloud
Version: 0.8.0
Summary: API Client library for Google Cloud
Home-page: https://github.com/GoogleCloudPlatform/gcloud-python
Author: Google Cloud Platform
Author-email: jjg+gcloud-python@google.com
License: Apache 2.0
Location: /usr/local/lib/python2.7/dist-packages
Requires: six, httplib2, pycrypto, protobuf, google-apitools, oauth2client
```

**Comment by dhermes** [comment]
@icrtiou Also thanks a lot for kicking the tires here **AND** for taking the time to report the bugs.

**Comment by dhermes** [comment]
Sorry the doc is vague:

```
:type rows: list of tuples
:param rows: row data to be inserted
```

As an example our [system test][1] uses rows as a list of triples that correspond to the [schema][2] used

```python
full_name = bigquery.SchemaField('full_name', 'STRING',
                                 mode='REQUIRED')
age = bigquery.SchemaField('age', 'INTEGER', mode='REQUIRED')
now = bigquery.SchemaField('now', 'TIMESTAMP')
```

to [create the table][3]

```python
table = dataset.table(TABLE_NAME, schema=[full_name, age, now])
```

(For the record, I wouldn't have expected you to have looked here, just wanted to provide an example.)

------

1. As of right now, we expect a `datetime.datetime` object for a `TIMESTAMP` schema field.
1. What situation are you in that you have lots of datetime's as strings rather than native? (So I can understand the use-case)
1. Could you provide the code snippets you used to generate the error? It's unclear how `sql_schema` and your "test data" were used.

[1]: https://github.com/GoogleCloudPlatform/gcloud-python/blob/0c74a599b6aa0a16e21575998e1770182c05e12c/system_tests/bigquery.py#L211-L216
[2]: https://github.com/GoogleCloudPlatform/gcloud-python/blob/0c74a599b6aa0a16e21575998e1770182c05e12c/system_tests/bigquery.py#L223-L226
[3]: https://github.com/GoogleCloudPlatform/gcloud-python/blob/0c74a599b6aa0a16e21575998e1770182c05e12c/system_tests/bigquery.py#L227

**Comment by icrtiou** [comment]
Perfect!! Thanks for quick response.
I'm on 0.8.0 doc. And

```
☁  pageview_to_bigquery [master] ⚡ pip show gcloud
---
Metadata-Version: 2.0
Name: gcloud
Version: 0.8.0
Summary: API Client library for Google Cloud
Home-page: https://github.com/GoogleCloudPlatform/gcloud-python
Author: Google Cloud Platform
Author-email: jjg+gcloud-python@google.com
License: Apache 2.0
Location: /Users/lucas/anaconda/lib/python2.7/site-packages
Requires: six, httplib2, pycrypto, protobuf, google-apitools, oauth2client
```

Thanks you so much by pointing out a good source to learn how to use gcloud API. The `system_tests` is wonderful.

The use case is as follow:
I got csv files periodically uploaded to s3, and I want to load files into bigquery table.
So I'm going to have many time string (UTC) like this `2015-12-14 18:29:01.245237` in the raw csv waiting to be loaded into bigquery.

About my code, it's very simple:
load data as a pandas dataframe -> transform it to json -> send it to bigquery table

`df = pd.read_csv(StringIO(data), names=['viewer_id', 'ip', 'author_id', 'post_id', 'time'])`
<img width="760" alt="screen shot 2016-01-11 at 9 29 50 pm" src="https://cloud.githubusercontent.com/assets/1643232/12255617/82387526-b8aa-11e5-9a12-1ac8603f94ad.png">

`rows = json.loads(df.to_json(orient='records'))`
<img width="472" alt="screen shot 2016-01-11 at 9 29 56 pm" src="https://cloud.githubusercontent.com/assets/1643232/12255626/9cfeb208-b8aa-11e5-9828-47c2c4f8007b.png">


`table.insert_data(rows)`

I use the `sql_schema` to create the bigquery table:
```
sql_schema = [('viewer_id', 'INTEGER'),
              ('ip','STRING'),
              ('author_id', 'INTEGER'),
              ('post_id', 'STRING'),
              ('time', 'TIMESTAMP')]

table.schema = [ SchemaField(name = n, field_type = t) for n, t in sql_schema ]
table.update()
```

So I guess I have to transform those raw time string into datetime object before loading into bigquery.
Thank you.

**Comment by dhermes** [comment]
@icrtiou You can upload your CSV files directly to Google Cloud Storage and then tell BigQuery to grab them ([ref][1]).

With our library you can use [`bigquery.Client.load_table_from_storage`][2] to accomplish this. See [storage docs][3] for info on how to upload the CSV to Google Cloud Storage (there is also an example in the BigQuery system test file).

[1]: https://cloud.google.com/bigquery/preparing-data-for-bigquery
[2]: http://googlecloudplatform.github.io/gcloud-python/0.8.0/bigquery-usage.html#inserting-data-asynchronous
[3]: http://googlecloudplatform.github.io/gcloud-python/0.8.0/storage-client.html

**Comment by icrtiou** [comment]
Got it. That's next I wanna try after I could handle this streaming case without leaving side effects. 
Thanks for your time.

**Comment by dhermes** [comment]
Thanks for your time! I am filing a new issue for consideration of how to handle existing string timestamps.

**Comment by icrtiou** [comment]
That would be awesome feature to have. For lazy library user XD

**Comment by dhermes** [comment]
Not just for lazy people but for people who don't want to waste compute cycles.

**Comment by icrtiou** [comment]
Indeed.

**Comment by jasonmellone** [comment]
Is there any reason table.insert() does not accept dictionaries of {'column_name':value_to_be_inserted,...,}?

This would be clean and allow arbitrary sorting.

**Comment by dhermes** [comment]
@jasonmellone Care to file a feature request?

**Comment by jasonmellone** [comment]
Will do this evening

**Comment by dhermes** [comment]
Thanks
